File: scripts/stock-workbench-control.py
#!/usr/bin/env python3
"""Drive the dashboard's ephemeral stock-workbench presentation state.

This tool changes only what the research dashboard displays. It cannot edit
research evidence, trade plans, broker holdings, orders, or database rows.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any


DEFAULT_BASE_URL = "http://127.0.0.1:5680"
DEFAULT_ENV_FILE = Path(r"G:\StockPlatform\config\runtime.env")
PANEL_NAMES = ("price", "metric", "next_session", "next_week", "messages", "trade_plan")


def read_env_file(path: Path) -> dict[str, str]:
    values: dict[str, str] = {}
    if not path.is_file():
        return values
    for raw_line in path.read_text(encoding="utf-8-sig").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        if line.startswith("export "):
            line = line[7:].lstrip()
        key, value = line.split("=", 1)
        value = value.strip()
        if len(value) >= 2 and value[0] == value[-1] and value[0] in "\"'":
            value = value[1:-1]
        values[key.strip()] = value
    return values


def request_json(base_url: str, path: str, *, payload: dict[str, Any] | None, operator_key: str) -> dict[str, Any]:
    body = None if payload is None else json.dumps(payload, ensure_ascii=False).encode("utf-8")
    headers = {"Accept": "application/json"}
    if body is not None:
        headers["Content-Type"] = "application/json"
    if operator_key:
        headers["X-Dashboard-Key"] = operator_key
    request = urllib.request.Request(
        f"{base_url.rstrip('/')}{path}", data=body, headers=headers,
        method="GET" if payload is None else "POST",
    )
    opener = urllib.request.build_opener(urllib.request.ProxyHandler({}))
    try:
        with opener.open(request, timeout=15) as response:
            return json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as error:
        response_text = error.read().decode("utf-8", errors="replace")
        try:
            message = json.loads(response_text).get("message") or response_text
        except json.JSONDecodeError:
            message = response_text
        raise RuntimeError(f"dashboard returned HTTP {error.code}: {message}") from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"dashboard is unavailable at {base_url}: {error.reason}") from error


def add_common_options(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("--base-url", default=os.getenv("STOCK_DASHBOARD_URL", DEFAULT_BASE_URL))
    parser.add_argument("--env-file", type=Path, default=DEFAULT_ENV_FILE)
    parser.add_argument("--workspace-id", default="primary")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Control the stock research workbench presentation")
    add_common_options(parser)
    subparsers = parser.add_subparsers(dest="command", required=True)

    subparsers.add_parser("show", help="read the current presentation state")

    present = subparsers.add_parser("present", help="open or update a stock presentation")
    present.add_argument("--symbol")
    present.add_argument("--lookback-days", type=int)
    present.add_argument("--strategy", choices=(
        "accumulation", "expansion", "pullback", "trend", "event", "relay",
        "contraction", "rotation", "reclaim",
    ))
    present.add_argument("--timeframe", choices=("daily", "weekly"))
    present.add_argument("--metric", choices=("vendor_flow", "volume", "turnover", "macd", "rsi"))
    present.add_argument("--zoom-start", type=float)
    present.add_argument("--zoom-end", type=float)
    present.add_argument("--focus-scope", choices=("next_session", "next_week"))
    present.add_argument("--focus-state")
    present.add_argument("--clear-focus", action="store_true")
    present.add_argument("--show-panel", action="append", choices=PANEL_NAMES, default=[])
    present.add_argument("--hide-panel", action="append", choices=PANEL_NAMES, default=[])
    present.add_argument("--note-title")
    present.add_argument("--note-body")
    present.add_argument("--note-source")
    present.add_argument("--note-as-of")
    present.add_argument("--clear-note", action="store_true")
    present.add_argument("--ttl-seconds", type=int, default=1800)

    annotate = subparsers.add_parser("annotate", help="add or replace a chart/narration annotation")
    annotate.add_argument("--kind", required=True, choices=("price_line", "point", "region", "note"))
    annotate.add_argument("--id")
    annotate.add_argument("--label", required=True)
    annotate.add_argument("--detail")
    annotate.add_argument("--color")
    annotate.add_argument("--source")
    annotate.add_argument("--as-of")
    annotate.add_argument("--price", type=float)
    annotate.add_argument("--date")
    annotate.add_argument("--start-date")
    annotate.add_argument("--end-date")
    annotate.add_argument("--low", type=float)
    annotate.add_argument("--high", type=float)
    annotate.add_argument("--ttl-seconds", type=int, default=1800)

    remove = subparsers.add_parser("remove", help="remove one annotation by id")
    remove.add_argument("annotation_id")
    remove.add_argument("--ttl-seconds", type=int, default=1800)
    subparsers.add_parser("clear", help="clear every agent annotation")
    subparsers.add_parser("reset", help="clear the entire presentation state")
    return parser


def compact(mapping: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in mapping.items() if value is not None}


def command_payload(args: argparse.Namespace) -> dict[str, Any] | None:
    if args.command == "show":
        return None
    common = {"workspace_id": args.workspace_id}
    if args.command == "reset":
        return {**common, "operation": "reset"}
    if args.command == "clear":
        return {**common, "operation": "clear_annotations"}
    if args.command == "remove":
        return {**common, "operation": "remove_annotation", "annotation_id": args.annotation_id, "ttl_seconds": args.ttl_seconds}
    if args.command == "annotate":
        annotation = compact({
            "kind": args.kind, "id": args.id, "label": args.label, "detail": args.detail,
            "color": args.color, "source_label": args.source, "as_of": args.as_of,
            "price": args.price, "date": args.date, "start_date": args.start_date,
            "end_date": args.end_date, "low": args.low, "high": args.high,
        })
        return {**common, "operation": "annotate", "annotation": annotation, "ttl_seconds": args.ttl_seconds}

    panels = {panel: True for panel in args.show_panel}
    panels.update({panel: False for panel in args.hide_panel})
    if args.clear_focus:
        focus: dict[str, str] | None = None
    elif args.focus_scope or args.focus_state:
        if not args.focus_scope or not args.focus_state:
            raise RuntimeError("--focus-scope and --focus-state must be supplied together")
        focus = {"scope": args.focus_scope, "state": args.focus_state}
    else:
        focus = None
    payload = compact({
        **common,
        "operation": "present",
        "symbol": args.symbol,
        "lookback_days": args.lookback_days,
        "strategy_key": args.strategy,
        "timeframe": args.timeframe,
        "metric": args.metric,
        "zoom": compact({"start": args.zoom_start, "end": args.zoom_end}) or None,
        "panel_visibility": panels or None,
        "ttl_seconds": args.ttl_seconds,
    })
    if args.clear_focus or focus is not None:
        payload["focus"] = focus
    if args.clear_note:
        payload["speaker_note"] = None
    elif args.note_title or args.note_body:
        if not args.note_title or not args.note_body:
            raise RuntimeError("--note-title and --note-body must be supplied together")
        payload["speaker_note"] = compact({
            "title": args.note_title, "body": args.note_body,
            "source_label": args.note_source, "as_of": args.note_as_of,
        })
    return payload


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    env_values = read_env_file(args.env_file)
    operator_key = os.getenv("DASHBOARD_OPERATOR_KEY", "") or env_values.get("DASHBOARD_OPERATOR_KEY", "")
    try:
        payload = command_payload(args)
        result = request_json(args.base_url, "/api/research/workbench-control", payload=payload, operator_key=operator_key)
    except RuntimeError as error:
        print(str(error), file=sys.stderr)
        return 1
    print(json.dumps(result, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: feishu-adapter/stock-workbench-control.mjs
import { randomUUID } from 'node:crypto';

const STRATEGIES = new Set([
	'accumulation', 'expansion', 'pullback', 'trend', 'event', 'relay',
	'contraction', 'rotation', 'reclaim',
]);
const TIMEFRAMES = new Set(['daily', 'weekly']);
const METRICS = new Set(['vendor_flow', 'volume', 'turnover', 'macd', 'rsi']);
const PANELS = new Set(['price', 'metric', 'next_session', 'next_week', 'messages', 'trade_plan']);
const ANNOTATION_KINDS = new Set(['price_line', 'point', 'region', 'note']);
const FOCUS_SCOPES = new Set(['next_session', 'next_week']);
const SYMBOL_PATTERN = /^\d{6}\.(?:SH|SZ|BJ)$/;
const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

function fail(message) {
	const error = new Error(message);
	error.statusCode = 400;
	throw error;
}

function finiteNumber(value, label, { min = -Infinity, max = Infinity, required = false } = {}) {
	if (value === undefined || value === null || value === '') {
		if (required) fail(`${label} is required`);
		return undefined;
	}
	const number = Number(value);
	if (!Number.isFinite(number) || number < min || number > max) fail(`${label} must be between ${min} and ${max}`);
	return number;
}

function shortText(value, label, maxLength, { required = false } = {}) {
	const text = String(value ?? '').trim();
	if (required && !text) fail(`${label} is required`);
	if (text.length > maxLength) fail(`${label} must be at most ${maxLength} characters`);
	return text || undefined;
}

function isoDate(value, label, { required = false } = {}) {
	const text = shortText(value, label, 10, { required });
	if (text && !DATE_PATTERN.test(text)) fail(`${label} must use YYYY-MM-DD`);
	return text;
}

function enumValue(value, allowed, label) {
	if (value === undefined || value === null || value === '') return undefined;
	const text = String(value).trim();
	if (!allowed.has(text)) fail(`${label} is not supported`);
	return text;
}

function clone(value) {
	return structuredClone(value);
}

function normalizeAnnotation(raw, idFactory) {
	if (!raw || typeof raw !== 'object' || Array.isArray(raw)) fail('annotation must be an object');
	const kind = enumValue(raw.kind, ANNOTATION_KINDS, 'annotation.kind');
	if (!kind) fail('annotation.kind is required');
	const annotation = {
		id: shortText(raw.id, 'annotation.id', 80) ?? idFactory(),
		kind,
		label: shortText(raw.label, 'annotation.label', 80, { required: true }),
		detail: shortText(raw.detail, 'annotation.detail', 600),
		color: shortText(raw.color, 'annotation.color', 32),
		source_label: shortText(raw.source_label, 'annotation.source_label', 120),
		as_of: shortText(raw.as_of, 'annotation.as_of', 40),
	};
	if (kind === 'price_line') annotation.price = finiteNumber(raw.price, 'annotation.price', { min: 0, required: true });
	if (kind === 'point') {
		annotation.date = isoDate(raw.date, 'annotation.date', { required: true });
		annotation.price = finiteNumber(raw.price, 'annotation.price', { min: 0, required: true });
	}
	if (kind === 'region') {
		annotation.start_date = isoDate(raw.start_date, 'annotation.start_date', { required: true });
		annotation.end_date = isoDate(raw.end_date, 'annotation.end_date', { required: true });
		annotation.low = finiteNumber(raw.low, 'annotation.low', { min: 0, required: true });
		annotation.high = finiteNumber(raw.high, 'annotation.high', { min: 0, required: true });
		if (annotation.start_date > annotation.end_date) fail('annotation.start_date must not be after end_date');
		if (annotation.low > annotation.high) fail('annotation.low must not exceed high');
	}
	return annotation;
}

function normalizeSpeakerNote(raw) {
	if (raw === null) return null;
	if (!raw || typeof raw !== 'object' || Array.isArray(raw)) fail('speaker_note must be an object or null');
	return {
		title: shortText(raw.title, 'speaker_note.title', 100, { required: true }),
		body: shortText(raw.body, 'speaker_note.body', 1200, { required: true }),
		source_label: shortText(raw.source_label, 'speaker_note.source_label', 120),
		as_of: shortText(raw.as_of, 'speaker_note.as_of', 40),
	};
}

function idleState(revision, now) {
	return {
		contract_version: 'stock-workbench-control.v1',
		revision,
		active: false,
		workspace_id: 'primary',
		updated_at: new Date(now).toISOString(),
		expires_at: null,
		symbol: null,
		lookback_days: 120,
		strategy_key: null,
		timeframe: 'daily',
		metric: null,
		zoom: { start: 55, end: 100 },
		focus: null,
		panel_visibility: Object.fromEntries([...PANELS].map((key) => [key, true])),
		annotations: [],
		speaker_note: null,
		presentation_only: true,
	};
}

export function createStockWorkbenchControl({ now = () => Date.now(), idFactory = randomUUID, defaultTtlSeconds = 1800 } = {}) {
	let state = idleState(0, now());

	function snapshot() {
		if (state.active && state.expires_at && Date.parse(state.expires_at) <= now()) {
			state = idleState(state.revision + 1, now());
		}
		return clone(state);
	}

	function apply(raw = {}) {
		if (!raw || typeof raw !== 'object' || Array.isArray(raw)) fail('control command must be an object');
		const operation = String(raw.operation ?? 'present').trim();
		if (!new Set(['present', 'annotate', 'remove_annotation', 'clear_annotations', 'reset']).has(operation)) fail('operation is not supported');
		const current = snapshot();
		if (operation === 'reset') {
			state = idleState(current.revision + 1, now());
			return snapshot();
		}

		const ttlSeconds = finiteNumber(raw.ttl_seconds, 'ttl_seconds', { min: 30, max: 86_400 }) ?? defaultTtlSeconds;
		const next = clone(current);
		next.revision += 1;
		next.active = true;
		next.workspace_id = shortText(raw.workspace_id ?? next.workspace_id, 'workspace_id', 64, { required: true });
		if (!/^[a-zA-Z0-9_-]+$/.test(next.workspace_id)) fail('workspace_id may only contain letters, digits, underscores and hyphens');
		next.updated_at = new Date(now()).toISOString();
		next.expires_at = new Date(now() + ttlSeconds * 1000).toISOString();

		if (operation === 'present') {
			if (raw.symbol !== undefined) {
				const symbol = String(raw.symbol ?? '').trim().toUpperCase();
				if (!SYMBOL_PATTERN.test(symbol)) fail('symbol must look like 600487.SH');
				next.symbol = symbol;
			}
			const lookback = finiteNumber(raw.lookback_days, 'lookback_days', { min: 30, max: 300 });
			if (lookback !== undefined) next.lookback_days = Math.trunc(lookback);
			const strategy = enumValue(raw.strategy_key, STRATEGIES, 'strategy_key');
			if (strategy) next.strategy_key = strategy;
			const timeframe = enumValue(raw.timeframe, TIMEFRAMES, 'timeframe');
			if (timeframe) next.timeframe = timeframe;
			const metric = enumValue(raw.metric, METRICS, 'metric');
			if (metric) next.metric = metric;
			if (raw.zoom !== undefined) {
				if (!raw.zoom || typeof raw.zoom !== 'object' || Array.isArray(raw.zoom)) fail('zoom must be an object');
				const start = finiteNumber(raw.zoom.start, 'zoom.start', { min: 0, max: 99 }) ?? next.zoom.start;
				const end = finiteNumber(raw.zoom.end, 'zoom.end', { min: 1, max: 100 }) ?? next.zoom.end;
				if (start >= end) fail('zoom.start must be below zoom.end');
				next.zoom = { start, end };
			}
			if (raw.focus === null) next.focus = null;
			else if (raw.focus !== undefined) {
				if (!raw.focus || typeof raw.focus !== 'object' || Array.isArray(raw.focus)) fail('focus must be an object or null');
				next.focus = {
					scope: enumValue(raw.focus.scope, FOCUS_SCOPES, 'focus.scope'),
					state: shortText(raw.focus.state, 'focus.state', 80, { required: true }),
				};
				if (!next.focus.scope) fail('focus.scope is required');
			}
			if (raw.panel_visibility !== undefined) {
				if (!raw.panel_visibility || typeof raw.panel_visibility !== 'object' || Array.isArray(raw.panel_visibility)) fail('panel_visibility must be an object');
				for (const [panel, visible] of Object.entries(raw.panel_visibility)) {
					if (!PANELS.has(panel)) fail(`unsupported panel: ${panel}`);
					if (typeof visible !== 'boolean') fail(`panel_visibility.${panel} must be boolean`);
					next.panel_visibility[panel] = visible;
				}
			}
			if (raw.speaker_note !== undefined) next.speaker_note = normalizeSpeakerNote(raw.speaker_note);
		}

		if (operation === 'annotate') {
			const annotation = normalizeAnnotation(raw.annotation, idFactory);
			next.annotations = [...next.annotations.filter((item) => item.id !== annotation.id), annotation].slice(-40);
		}
		if (operation === 'remove_annotation') {
			const id = shortText(raw.annotation_id, 'annotation_id', 80, { required: true });
			next.annotations = next.annotations.filter((item) => item.id !== id);
		}
		if (operation === 'clear_annotations') next.annotations = [];
		state = next;
		return snapshot();
	}

	return { snapshot, apply };
}

export const stockWorkbenchControlVocabulary = Object.freeze({
	strategies: [...STRATEGIES], timeframes: [...TIMEFRAMES], metrics: [...METRICS], panels: [...PANELS], annotation_kinds: [...ANNOTATION_KINDS], focus_scopes: [...FOCUS_SCOPES],
});
